File: typescript/attestations/test_examples.ts
/**
 * Test script to verify attestation example output consistency.
 *
 * Tests three scenarios:
 * 1. Unknown event (404)
 * 2. Unresolved event (finalised_time=null, outcome=null)
 * 3. Resolved event (finalised_time=timestamp, outcome=data)
 */

import type { AttestationResponse } from "./types.js";
import { getJson } from "./client.js";

async function testUnknownEvent(): Promise<string> {
  const eventId = "evt_unknown";
  const outputLines: string[] = [];
  outputLines.push(`Fetching attestation for event: ${eventId}`);
  outputLines.push("-".repeat(50));

  try {
    await getJson<AttestationResponse>(`/api/v1/public/attestations/${eventId}`);
    outputLines.push("ERROR: Expected 404 but request succeeded");
  } catch (err) {
    const error = err as Error;
    if (error.message.includes("404")) {
      outputLines.push("Status: UNKNOWN");
      outputLines.push(`Event '${eventId}' not found (404)`);
      outputLines.push("The event_id does not exist in the database.");
    } else {
      outputLines.push(`ERROR: Unexpected HTTP error: ${error.message}`);
    }
  }

  return outputLines.join("\n");
}

async function testUnresolvedEvent(): Promise<string> {
  const eventId = "evt_test_123";
  const outputLines: string[] = [];
  outputLines.push(`Fetching attestation for event: ${eventId}`);
  outputLines.push("-".repeat(50));

  try {
    const attestation: AttestationResponse = await getJson<AttestationResponse>(
      `/api/v1/public/attestations/${eventId}`,
    );
    outputLines.push(`Event ID: ${attestation.event_id}`);
    outputLines.push(`Start Time: ${attestation.start_time}`);
    outputLines.push(`End Time: ${attestation.end_time}`);

    const { finalised_time: finalisedTime, outcome } = attestation;

    if (finalisedTime === null && outcome === null) {
      outputLines.push("");
      outputLines.push("Status: UNRESOLVED");
      outputLines.push("The event exists but has not been finalised yet.");
    } else {
      outputLines.push("");
      outputLines.push("ERROR: Expected unresolved but got different state");
    }
  } catch (err) {
    outputLines.push(`ERROR: Unexpected HTTP error: ${(err as Error).message}`);
  }

  return outputLines.join("\n");
}

async function testResolvedEvent(): Promise<string> {
  const eventId = "evt_resolved_456";
  const outputLines: string[] = [];
  outputLines.push(`Fetching attestation for event: ${eventId}`);
  outputLines.push("-".repeat(50));

  try {
    const attestation: AttestationResponse = await getJson<AttestationResponse>(
      `/api/v1/public/attestations/${eventId}`,
    );
    outputLines.push(`Event ID: ${attestation.event_id}`);
    outputLines.push(`Start Time: ${attestation.start_time}`);
    outputLines.push(`End Time: ${attestation.end_time}`);

    const { finalised_time: finalisedTime, outcome } = attestation;

    if (finalisedTime !== null && outcome !== null) {
      outputLines.push("");
      outputLines.push("Status: RESOLVED");
      outputLines.push(`Finalised at: ${finalisedTime}`);
      outputLines.push("Outcome:");
      outputLines.push(JSON.stringify(outcome, null, 2));
    } else {
      outputLines.push("");
      outputLines.push("ERROR: Expected resolved but got different state");
    }
  } catch (err) {
    outputLines.push(`ERROR: Unexpected HTTP error: ${(err as Error).message}`);
  }

  return outputLines.join("\n");
}

async function main(): Promise<void> {
  console.log("=".repeat(60));
  console.log("TYPESCRIPT ATTESTATION TESTS");
  console.log("=".repeat(60));

  console.log("\n--- Test 1: Unknown Event (404) ---\n");
  console.log(await testUnknownEvent());

  console.log("\n--- Test 2: Unresolved Event ---\n");
  console.log(await testUnresolvedEvent());

  console.log("\n--- Test 3: Resolved Event ---\n");
  console.log(await testResolvedEvent());

  console.log("\n" + "=".repeat(60));
  console.log("ALL TESTS COMPLETE");
  console.log("=".repeat(60));
}

main().catch((err: unknown) => {
  console.error(err);
  process.exit(1);
});

File: python/attestations/test_examples.py
"""
Test script to verify attestation example output consistency.

Tests three scenarios:
1. Unknown event (404)
2. Unresolved event (finalised_time=null, outcome=null)
3. Resolved event (finalised_time=timestamp, outcome=data)
"""

from __future__ import annotations

import json
import os
from typing import Any

import requests


def _base_url() -> str:
    base = os.environ.get("ANERA_MARKETS_API_BASE_URL", "https://api.anera.markets").strip().rstrip("/")
    return base


def get_attestation(event_id: str) -> dict[str, Any]:
    url = f"{_base_url()}/api/v1/public/attestations/{event_id}"
    r = requests.get(url, timeout=60)
    r.raise_for_status()
    return r.json()


def test_unknown_event() -> str:
    """Test unknown event returns 404."""
    event_id = "evt_unknown"
    output_lines = []
    output_lines.append(f"Fetching attestation for event: {event_id}")
    output_lines.append("-" * 50)

    try:
        get_attestation(event_id)
        output_lines.append("ERROR: Expected 404 but request succeeded")
    except requests.HTTPError as e:
        if e.response is not None and e.response.status_code == 404:
            output_lines.append("Status: UNKNOWN")
            output_lines.append(f"Event '{event_id}' not found (404)")
            output_lines.append("The event_id does not exist in the database.")
        else:
            output_lines.append(f"ERROR: Unexpected HTTP error: {e}")

    return "\n".join(output_lines)


def test_unresolved_event() -> str:
    """Test unresolved event returns null finalised_time and outcome."""
    event_id = "evt_test_123"
    output_lines = []
    output_lines.append(f"Fetching attestation for event: {event_id}")
    output_lines.append("-" * 50)

    try:
        attestation = get_attestation(event_id)
        output_lines.append(f"Event ID: {attestation['event_id']}")
        output_lines.append(f"Start Time: {attestation['start_time']}")
        output_lines.append(f"End Time: {attestation['end_time']}")

        finalised_time = attestation.get("finalised_time")
        outcome = attestation.get("outcome")

        if finalised_time is None and outcome is None:
            output_lines.append("")
            output_lines.append("Status: UNRESOLVED")
            output_lines.append("The event exists but has not been finalised yet.")
        else:
            output_lines.append("")
            output_lines.append("ERROR: Expected unresolved but got different state")

    except requests.HTTPError as e:
        output_lines.append(f"ERROR: Unexpected HTTP error: {e}")

    return "\n".join(output_lines)


def test_resolved_event() -> str:
    """Test resolved event returns non-null finalised_time and outcome."""
    event_id = "evt_resolved_456"
    output_lines = []
    output_lines.append(f"Fetching attestation for event: {event_id}")
    output_lines.append("-" * 50)

    try:
        attestation = get_attestation(event_id)
        output_lines.append(f"Event ID: {attestation['event_id']}")
        output_lines.append(f"Start Time: {attestation['start_time']}")
        output_lines.append(f"End Time: {attestation['end_time']}")

        finalised_time = attestation.get("finalised_time")
        outcome = attestation.get("outcome")

        if finalised_time is not None and outcome is not None:
            output_lines.append("")
            output_lines.append("Status: RESOLVED")
            output_lines.append(f"Finalised at: {finalised_time}")
            output_lines.append("Outcome:")
            output_lines.append(json.dumps(outcome, indent=2))
        else:
            output_lines.append("")
            output_lines.append("ERROR: Expected resolved but got different state")

    except requests.HTTPError as e:
        output_lines.append(f"ERROR: Unexpected HTTP error: {e}")

    return "\n".join(output_lines)


def main() -> None:
    print("=" * 60)
    print("PYTHON ATTESTATION TESTS")
    print("=" * 60)

    print("\n--- Test 1: Unknown Event (404) ---\n")
    print(test_unknown_event())

    print("\n--- Test 2: Unresolved Event ---\n")
    print(test_unresolved_event())

    print("\n--- Test 3: Resolved Event ---\n")
    print(test_resolved_event())

    print("\n" + "=" * 60)
    print("ALL TESTS COMPLETE")
    print("=" * 60)


if __name__ == "__main__":
    main()
